### PR TITLE
Add octave build commands

### PR DESCRIPTION
**Pull request by matsievskiysv**
Add support for build with octave.
Some functions do not work with octave though. More work is needed. Maybe this could be pulled into separate branch?

### CODE REVIEW

**Comment by lfarv** [comment]
Hello @matsievskiysv,
I have a few comments on your proposals:
- Concerning the build commands: can you try to avoid makefiles ? These are not portable across architectures, and may need some tuning by the user. Using the `mkoctfile --mex`  or `mex` command within Octave should provide a universal way of building extensions, including on Windows. Is there any problem with that (I am not familiar with Octave)? Also, rather than modifying `atmexall` (already complicated enough), you can provide a dedicated build function for Octave.
- About missing Matlab functions (ex.: `endsWith`): one could add a top-level directory (`atoctave` for instance). It should be ignored by Matlab users (only `atmat` and `atintegrators` in the path), and should be put in 1st position by Octave users. There, you could put any function that simulates a missing Matlab function, or that replaces an `atmat` function for Octave users. Then you can keep the name `endsWith` (for instance) wherever it appears and provide your own version in `atoctave`.
- You can forget the files in the `atintegrators/user` directory: they are still there for historical reasons, by have not been upgraded for the python interface and are not compiled!

Finally, I agree with you idea of temporarily creating a dedicated branch for octave, until it can be merged into the main one.

**Comment by matsievskiysv** [comment]
@lfarv
> Concerning the build commands: can you try to avoid makefiles ...
  
Replaced with file `bootstrapOctave.m`

> About missing Matlab functions ...
  
Done

> You can forget the files ...

Removed them from source list

> Finally, I agree with you idea of temporarily creating a dedicated branch for octave, until it can be merged into the main one.

As soon as you'll create an `octave` branch, I'll set it as this PR's merge target.

**Comment by lfarv** [comment]
There is now a branch `octave_compatibility`. It adds an `atoctave` directory, where you should put all octave-specific .m files, so that they are ignored in Matlab.

**Comment by matsievskiysv** [comment]
Ok. Demos seems to work except for the GUI stuff and ones that fail because of #200

**Comment by matsievskiysv** [comment]
I don't know how the output of the demos should look like, so I attach graphical output:

---

**atfastring/testatfastring.m**

fastring-1
fastring-2

---

**atmatchExamples/bump/run_BumpFit.m**

bump

---

**atmatchExamples/ExampleATMATCH/runtotest_atmatch.m**

exampleATMATCH

---

**IDModeling/trackWithU35KickMap.m**

trackWithU35KickMap-1

trackWithU35KickMap-2

---

**opticsAndBeamSizes/linoptdemo.m**

*This one is obviously messed up*

linpotdemo

---

**opticsAndBeamSizes/ohmienvelopedemo.m**

ohmienvelopedemo-1

ohmienvelopedemo-2

---

**responseMatrix/findrespmdemo.m**

findrespmdemo-1

findrespmdemo-2

**Comment by lfarv** [comment]
Hi @matsievskiysv, That's excellent ! There are still a few problems, but it's already very promising. On the demos, there is a problem with H/V coupling: you pointed the problem on atinopt, bit on ohmienvelope, on the right axis, the scale should be in the 1E-4 while you get 1E-18. Also, in findrespmdemo, the off-diagonal blocks are at zero, which is normal for the non-coupled ring, but not for the coupled ring. Finally, a detail, on some plots (bump fit, attach), you should get a synoptic representation of the ring magnets along the horizontal axis. I'll think about what could cause the tracking problems for could machines: it's like the tilted magnets would not couple horizontal position into the vertical plane…

**Comment by lfarv** [comment]
Sorry to intervene again:

- For compilation of integrators: you should select  all ‘*Pass.c’ and ‘*Pass.cc’ files only. All other *.c or *.cc files are meant to be ‘#include’d in integrators, but not compiled independently.
- Can you avoid modifying all unused ‘atintegrators/user/*.c files ? The less modified files, the easier the review…
- I am worried by the requirement for the un-released Octave version 6. In my experience, system administrators are reluctant to continuously upgrade software packages to the last version, and users have often to leave with slightly older packages. This could block many potential users. I don't know how many functions are affected by the bug you mention, but at least the most critical ones could be re-written in the `atoctave` directory by extracting the problematic nested functions. For others, a partially functioning AT could be better than nothing.

**Comment by lfarv** [comment]
For info: did you try `atdiag`? Here is the expected output:

test

**Comment by matsievskiysv** [comment]
@lfarv  About `function_handle` for nested functions. It seems that I can wrap them into anonymous function:
```octave
[x,y,z] = dostuff(@handle, arg)		% replace this
[x,y,z] = dostuff(@(a) handle(a), arg)	% with this
```
Is this acceptable?

**Comment by matsievskiysv** [comment]
By the way, Octave 6.1 [is out](https://www.gnu.org/software/octave/news/release/2020/11/26/octave-6.1.0-released.html)

**Comment by matsievskiysv** [comment]
I get this for `atdiag`:
diag

Plot legend and annotations are missing.

**Comment by matsievskiysv** [comment]
> For compilation of integrators: you should select all ‘*Pass.c’ and ‘*Pass.cc’ files only. All other *.c or *.cc files are meant to be ‘#include’d in integrators, but not compiled independently.

Edited

> Can you avoid modifying all unused ‘atintegrators/user/*.c files ? The less modified files, the easier the review…

Removed edits

> I am worried by the requirement for the un-released Octave version 6. In my experience, system administrators are reluctant to continuously upgrade software packages to the last version, and users have often to leave with slightly older packages. This could block many potential users. I don't know how many functions are affected by the bug you mention, but at least the most critical ones could be re-written in the atoctave directory by extracting the problematic nested functions. For others, a partially functioning AT could be better than nothing.

Unfortunately for now only 6.1 works. There's also some incompatibility in line continuations (`...`).

**Comment by matsievskiysv** [comment]
> after small alterations of atplotsyn I get this

It also helped with some other plots

bump
atmatch1
atmatch2
atmatch3

File: atmat/attrack/linepass.m [hunk 1]
```diff
@@ -90,7 +90,7 @@
 
 newlattice = double(~keeplattice);
 
-[prefunc,postfunc]=parseargs({function_handle.empty,function_handle.empty},...
+[prefunc,postfunc]=parseargs({cell(),cell()},...
```
Comment (lfarv) [inline atmat/attrack/linepass.m:93]: Works, but isn't there any way to create empty arrays on any type in Octave?

File: atmat/pubtools/lattice_tools/atreadbeta.m [hunk 2]
```diff
@@ -160,7 +160,7 @@
 end
 
 
-evalin('base','global GLOBVAL');
+assignin('base','GLOBVAL',GLOBVAL);
```
Comment (lfarv) [inline atmat/pubtools/lattice_tools/atreadbeta.m:163]: It is different from the original: here you create a local variable in the base workspace, identical to the global one (bad idea). Declaring the global variable in the base workspace makes it visible, but that's not so useful. The best is to suppress completely these statements (and the similar ones)

**Review by lfarv — COMMENTED** [review]
I do not see incompatibilities with Matlab. One could merge this into the octave branch and see!

**Comment by lfarv** [comment]
Now we are left with the problem of wrong results for coupled machine. I cannot get Octave 6.1 on MacOS, and I understood that you don't have access to Matlab. That makes checks difficult… The simpler check is just comparing numerical results for pure tracking over 1 turn of a coupled machine, then go to closed orbit, and then to transfer matrices. If you do not see an easier way, I could send you .mat files with Matlab results, can you read them? It may be slow since I cannot spend so much time on AT, but we can try.

**Comment by matsievskiysv** [comment]
> I could send you .mat files with Matlab results, can you read them

I [should be able to](https://stackoverflow.com/questions/9915658/how-do-you-open-mat-files-in-octave), if `.mat` versions will match.

Maybe I'll be able to extract some information from https://github.com/atcollab/ATtraining examples as well.

**Comment by matsievskiysv** [comment]
I've created a [docker](https://www.docker.com/) image with a [jupyter server](https://jupyter.org/) with [Octave](https://github.com/calysto/octave_kernel) and AT. Not all plots show, but it's sufficient for testing purposes.

@lfarv If you'll give me your email address, I could send you a link to my personal server so you could test AT there.

image

image

In order to use it:
1. Install [Docker](https://docs.docker.com/docker-for-mac/install/) (should be trivial)
1. Run command `docker run -d --name at -p 8888:8888 matsievskiysv/at`
1. Run `docker logs at --follow` and write down connection token (will appear on screen as `http://127.0.0.1:8888/?token=<token>`)
1. From web-browser go to `http://127.0.0.1:8888/` page and insert token there.
1. Create a notebook and execute commands. All AT should be available.

File: atmat/attrack/linepass.m (hunk 1 above)
Comment (matsievskiysv) [reply atmat/attrack/linepass.m:93]: There's no `function_handle` class in Octave. In any case, anything empty will work: https://github.com/atcollab/at/blob/master/atmat/attrack/atpass.c#L403-L404

File: atmat/pubtools/lattice_tools/atreadbeta.m (hunk 2 above)
Comment (matsievskiysv) [reply atmat/pubtools/lattice_tools/atreadbeta.m:163]: Original statement made Octave crush. If these are not needed, I will gladly remove them.

**Comment by willrogers** [comment]
This could be a really useful development, particularly allowing us to run Matlab tests and Python tests on pull requests and maybe compare the results. I'm very pleased someone is trying this, thanks!

I see Octave 6.1 was released on November 26th, but anyway since we have no Octave support at all at the moment I wouldn't be worried that it requires a new version.

File: atmat/pubtools/lattice_tools/atreadbeta.m (hunk 2 above)
Comment (lfarv) [reply atmat/pubtools/lattice_tools/atreadbeta.m:163]: Go on, remove all the similar lines that you pointed out

**Comment by lfarv** [comment]
@matsievskiysv:

- I'll try the 'docker' way as soon as I'm back home (next monday probably). How shall I get your image?. If that fails, I'll try the access to your server,
- I propose that you continue the development of the octave port directly on the `octave_compatibility` branch. It will remove one intermediate step.

**Comment by matsievskiysv** [comment]
> How shall I get your image?

`docker run -d --name at -p 8888:8888 matsievskiysv/at` command will download [it](https://hub.docker.com/r/matsievskiysv/at) automatically.

> I propose that you continue the development of the octave port directly on the octave_compatibility branch. It will remove one intermediate step.

It's fine by me

**Comment by lfarv** [comment]
Here is a set of 2 files for checking `atlinop` with octave:

1. Remove the .txt extension which I added to get them accepted as attachments and put them anywhere in the octave path
2. execute checklinopt with either;

- `[results,expected]=checklinopt('coupled');`
- `[results,expected]=checklinopt('uncoupled');`

`results` and `expected` are structures containing the output of `atlinopt` in `lindata`, `tunes`, `chrom` fields.
'coupled' and 'uncoupled' run the same ring (spear2 example) without and with H/V coupling, introduced with quadrupole rotations.

[checklinopt.m.txt](https://github.com/atcollab/at/files/5637601/checklinopt.m.txt)
[checkring.mat.txt](https://github.com/atcollab/at/files/5637602/checkring.mat.txt)

**Comment by matsievskiysv** [comment]
@lfarv Results look identical

**Comment by matsievskiysv** [comment]
Got jupyter plots working. They do not look as good as Qt plots unfortunately.

image

File: atmat/atmatch/atmatch.m [hunk 3]
```diff
@@ -131,7 +131,7 @@
 usepar = p.Results.UseParallel;
 
 
-options=optimset(minimizer);
+options=optimset(func2str(minimizer));
```
[the remark this answers is not in the document]
Comment (matsievskiysv) [reply atmat/atmatch/atmatch.m]: Done

File: atmat/pubtools/lattice_tools/atreadbeta.m (hunk 2 above)
Comment (matsievskiysv) [reply atmat/pubtools/lattice_tools/atreadbeta.m]: It broke some code. Particularly, `machine_data`. I'll need some time to fix it

**Comment by lfarv** [comment]
@matsievskiysv: you should now be allowed to merge you fork into this branch. Let me know

**Comment by matsievskiysv** [comment]
Yes, I can merge now. But I want to fix `evalin` issues first

**Comment by matsievskiysv** [comment]
Don't know exactly why, but response matrix for coupled lattice now looks like this:

image

**Comment by lfarv** [comment]
For `evalin` problems: some demos are using a obsolete way of keeping the 'active' ring structure in the global variable "`THERING`". Though this should still work, it is now recommended to set the ring structure as 1st argument of most functions.

For response matrix: it now looks good!

For `atlinopt`: fine if numerical results are good. I found out that the `linopt` demo take a non coupled ring (spear2), and introduces coupling by adding randomly generated tilts. This is not a good idea since the demo will give different results each time it is run!  But this does not explain why the coupling looked zero in you plot. Did you try again since the response matrix looks good?

**Comment by matsievskiysv** [comment]
`linoptdemo`:
image
`ohmienvelopedemo`:
image
image

**Comment by matsievskiysv** [comment]
Jupyter inline graphics looks a bit sad, but works.

image

**Pull request merged**